lib/_http_client.js

```diff
@@ -350,11 +361,26 @@ ClientRequest.prototype.abort = function abort() {
   // In the event that we don't have a socket, we will pop out of
   // the request queue through handling in onSocket.
   if (this.socket) {
-    // in-progress
-    this.socket.destroy();
+    _destroy(this, this.socket, err);
+  } else if (err) {
+    this[kError] = err;
   }
 };
 
+function _destroy(req, socket, err) {
+  if (!req.agent || req.headersSent) {
```

Comment:
This assumes that `!req.headersSent` means that the socket was not used.

Reply:
@szmarczak I had to revert this. It breaks some tests. I consider it a minor optimization. We could look into it in the future.

Reply:
Sure, just mark this as a TODO.

Reply:
Done